src/python/dxpy/nextflow/ImageRef.py

```diff
@@ -0,0 +1,140 @@
+        :param file_name: A file name under which the image will be saved on the platform
+        :type file_name: String
+        :returns: Tuple[String, String] dx file id, file name (basename)
+        """
+        raise NotImplementedError("Abstract class. Method not implemented. Use the concrete implementations.")
+
+    def _reconstruct_image_ref(self):
+        raise NotImplementedError("Abstract class. Method not implemented. Use the concrete implementations.")
+
+    def _construct_cache_file_name(self):
+        return self._join_if_exists("_", [self._image_name, self._tag])
+
+    @staticmethod
+    def _join_if_exists(delimiter, parts):
+        return delimiter.join([x for x in parts if x])
+
+    def _dx_file_get_name(self):
+        dx_file_handle = DXFile(self._dx_file_id, config["DX_PROJECT_CONTEXT_ID"])
+        return dx_file_handle.describe().get("name")
+
+    def _package_bundle(self):
+        """
+        Function to include a container image stored on the platform into NPA
+        :returns: Dict in the format of {"name": "bundle.tar.gz", "id": {"$dnanexus_link": "file-xxxx"}}
+        """
+        if not self._dx_file_id:
+            cache_file_name = self._construct_cache_file_name()
+            self._dx_file_id = self._cache(cache_file_name)
+        else:
+            cache_file_name = self._dx_file_get_name()
+        return {
+            "name": cache_file_name,
+            "id": {"$dnanexus_link": self._dx_file_id}
+        }
+
+
+class DockerImageRef(ImageRef):
+    def __init__(
+            self,
+            process,
+            digest,
+            dx_file_id=None,
+            repository=None,
+            image_name=None,
+            tag=None
+    ):
+        super().__init__(
+            process,
+            digest,
+            dx_file_id,
+            repository,
+            image_name,
+            tag)
+
+    def _cache(self, file_name):
+        full_image_ref = self._reconstruct_image_ref()
+        docker_pull_cmd = "docker pull {}".format(full_image_ref)
+        docker_save_cmd = "docker save {} | gzip > {}".format(full_image_ref, file_name)
+        for cmd in [docker_pull_cmd, docker_save_cmd]:
+            try:
+                _ = subprocess.check_output(cmd, shell=True)
+            except subprocess.CalledProcessError:
+                err_exit("Failed to run a subprocess command: {}".format(cmd))
+        # may need wait_on_close = True??
+        uploaded_dx_file = upload_local_file(
+            filename=file_name,
+            project=config["DX_PROJECT_CONTEXT_ID"],
+            folder=self._caching_dir,
+            name=file_name,
+            parents=True,
+            properties={"image_digest": self._digest}
+        )
+        return uploaded_dx_file.get_id()
+
+    def _reconstruct_image_ref(self):
+        """
+        Docker image reference has the form of <REPOSITORY_NAME>/<IMAGE_NAME>:<VERSION_TAG>
+        """
```

Comment:
The docstring should be updated as well?

Reply:
Sure, fixed in [3a84ddc](https://github.com/dnanexus/dx-toolkit/pull/1263/commits/3a84ddc6932d82affeca9622538a1a7524144a71)

Reply:
Thanks!